
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Usage &#8212; smalr 1.1 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributing" href="contributing.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="contributing.html" title="Contributing"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">smalr 1.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="important-notes">
<h2>Important notes<a class="headerlink" href="#important-notes" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>For best results, cmp.h5 or BAM files generated from short library sequencing should be paired with the SMsn protocol.</li>
<li>Conversely, cmp.h5 or BAM files containing long library-sequenced reads are best when using the SMp protocol.</li>
<li>A CCS fastq file is only <em>required</em> when the –align option is specified [DEPRECATED USAGE; see below]. This calls an alignment step that is used to mask out sequencing errors on each molecule. –align should never be used with the SMp protocol and long libraries, as CCS sequencing only works with short libraries.</li>
<li>UPDATE: because the –align option has been deprecated, just specify NONE as the fastq path in the input file.</li>
</ul>
</div>
<div class="section" id="example-usage">
<h2>Example usage<a class="headerlink" href="#example-usage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">Usage: smalr [--help] [options] &lt;inputs_file&gt;</span>

<span class="go">This program will take a native aligned_reads.cmph5 and compare each molecule&#39;s</span>
<span class="go">kinetics profile with a matching WGA aligned_reads.cmp.h5. There are two protocols</span>
<span class="go">available for use within the pipeline, SMsn and SMp. The single argument for both</span>
<span class="go">protocols is an inputs_file containing paths to the relevant files.</span>

<span class="go">The inputs_file has the following format:</span>

<span class="go">native_cmph5 : /path/to/native/aligned_reads.cmp.h5</span>
<span class="go">fastq        : /path/to/native/CCS_reads.fastq       (DEPRECATED; specify NONE)</span>
<span class="go">wga_cmph5    : /path/to/WGA/aligned_reads.cmp.h5</span>
<span class="go">ref          : /path/to/sample/reference.fasta</span>

<span class="go">With the updated support for aligned BAM files (in addition to *.cmp.h5 files),</span>
<span class="go">please supply the path to the aligned BAM file using the &#39;native_cmph5&#39; and</span>
<span class="go">&#39;wga_cmph5&#39; lines in the input file.</span>


<span class="go">SMsn: Single-molecule, single nucleotide analysis</span>
<span class="go">Each motif site on each sequencing molecule is assessed for methylation</span>
<span class="go">status. This is designed for use with short (~250bp) sequencing library</span>
<span class="go">preps, where the long read lengths of SMRT reads enables multiple passes</span>
<span class="go">over each motif site. The reliability of the SMsn scores increases with</span>
<span class="go">more passes (i.e. higher single-molecule coverage).</span>

<span class="go">Example usage for SMsn analysis:</span>
<span class="go">smalr -i --SMsn --motif=GATC --mod_pos=2 --procs=4 -c 5 input.txt</span>


<span class="go">SMp: Single-molecule, motif-pooled analysis</span>
<span class="go">All motif sites on a sequencing molecule are pooled together and the</span>
<span class="go">molecule-wide methylation status for the given motif is assessed. This</span>
<span class="go">is designed for use with long (10Kb+) sequencing library preps, where each</span>
<span class="go">single long subread can span many distinct motif sites. The reliability of</span>
<span class="go">the SMp scores increases with increasing number of distinct motif sites</span>
<span class="go">contained in the subread.</span>

<span class="go">Example usage for SMp analysis:</span>
<span class="go">smalr -i --SMp --motif=GATC --mod_pos=2 --procs=4 -c 5 input.txt</span>


<span class="go">Options:</span>
<span class="go">-h, --help                                              Show this help message and exit</span>
<span class="go">-d, --debug                                             Increase verbosity of logging</span>
<span class="go">-i, --info                                              Add basic logging</span>
<span class="go">--logFile=LOGFILE                                       Write logging to file [log.out]</span>
<span class="go">--out=OUT                                               Filename to output SMsn/SMp results [&lt;SMsn/SMp&gt;.out]</span>
<span class="go">-c NATIVECOVTHRESH, --nativeCovThresh=NATIVECOVTHRESH   Per mol/strand coverage threshold below which to ignore molecules [10]</span>
<span class="go">-m MOTIF, --motif=MOTIF                                 (Required) The sequence motif to be analyzed [None]</span>
<span class="go">-s MOD_POS, --mod_pos=MOD_POS                           (Required) The modified position in the motif to be analyzed (e.g. for</span>
<span class="go">                                                        Gm6ATC, mod_pos=2) [None]</span>
<span class="go">--wgaCovThresh=WGACOVTHRESH                             Aggregate WGA coverage threshold below which to skip analysis at that</span>
<span class="go">                                                        position [10]</span>
<span class="go">--SMsn                                                  Use short-library, single-nucleotide detection protocol. [False]</span>
<span class="go">--SMp                                                   Use long-library epigenetic phasing protocol (pool IPDs from each</span>
<span class="go">                                                        subread). [False]</span>
<span class="go">--procs=PROCS                                           Number of processors to use [4]</span>
<span class="go">--align                                                 [DEPRECATED; not supported for BAM input] Align native reads to</span>
<span class="go">                                                        reference to avoid real SNP positions. Only use when expecting sequence</span>
<span class="go">                                                        heterogeneity in sample (i.e. mtDNA). [False]</span>
<span class="go">--upstreamSkip=UPSTREAMSKIP                             Number of bases 5&#39; of a CCS-detected, molecule-level SNP to skip in</span>
<span class="go">                                                        analysis (only when using --align) [10]</span>
<span class="go">--downstreamSkip=DOWNSTREAMSKIP                         Number of bases 3&#39; of a CCS-detected, molecule-level SNP to skip in</span>
<span class="go">                                                        analysis (only when using --align) [10]</span>
<span class="go">--minSubreadLen=MINSUBREADLEN                           Minimum length of a subread to analyze [100]</span>
<span class="go">--minAcc=MINACC                                         Minimum accuracy of a subread to analyze [0.8]</span>
<span class="go">--natProcs=NATPROCS                                     Number of processors to use for native molecule analysis [same as procs]</span>
<span class="go">--leftAnchor=LEFTANCHOR                                 Number of left bp to exclude around subread-level alignment errors [1]</span>
<span class="go">--rightAnchor=RIGHTANCHOR                               Number of right bp to exclude around subread-level alignment errors [1]</span>
<span class="go">--write_vars=WRITE_VARS                                 Write mol-specific variant calls to this filename (only when using --align)</span>
<span class="go">                                                        [None]</span>
<span class="go">--useZMW                                                Index molecules using ZMW/movie ID&#39;s instead of molecule IDs (if all</span>
<span class="go">                                                        alignments all have unique molecule IDs) [False]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Because the <code class="docutils literal notranslate"><span class="pre">--align</span></code> option has been deprecated due to limited use-cases, several of the above options are no longer necessary or useful. These include the following options, which can be disregarded:</dt>
<dd><ul class="first last simple">
<li><code class="docutils literal notranslate"><span class="pre">--align</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">--upstreamSkip</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">--downstreamSkip</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">--leftAnchor</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">--rightAnchor</span></code></li>
<li><code class="docutils literal notranslate"><span class="pre">--write_vars</span></code></li>
</ul>
</dd>
</dl>
</div>
<div class="section" id="pipeline-output">
<h2>Pipeline output<a class="headerlink" href="#pipeline-output" title="Permalink to this headline">¶</a></h2>
<p>One output directory will be created for each contig in the reference. If there is only one contig, the results will be placed in the folder named for that contig. These results include a log detailing the analysis of that contig, the motif positions in that contig (forward and reverse strand), a fasta file of that contig, and a results file (SMsn.out or SMp.out). This results file contains the following informtation:</p>
<div class="section" id="column-headers">
<h3>Column  headers<a class="headerlink" href="#column-headers" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple">
<li>Contig strand</li>
<li>Contig motif position (for SMp, pooled motif sites are summarized by smallest site position)</li>
<li>SMsn or SMp score (native score - WGA score)</li>
<li>Molecule ID</li>
<li>Native score (mean of subread-normalized ln(IPD) values; site- and molecule-specific)</li>
<li>WGA score (mean of subread-normalized ln(IPD) values; site-specific accross all WGA molecules)</li>
<li>Number of data points used to get the molecule-level native score</li>
<li>Number of data points used to get the aggregate WGA score</li>
<li>Mean length of subreads from the native molecule</li>
</ol>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Usage</a><ul>
<li><a class="reference internal" href="#important-notes">Important notes</a></li>
<li><a class="reference internal" href="#example-usage">Example usage</a></li>
<li><a class="reference internal" href="#pipeline-output">Pipeline output</a><ul>
<li><a class="reference internal" href="#column-headers">Column  headers</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="installation.html"
                        title="previous chapter">Installation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="contributing.html"
                        title="next chapter">Contributing</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/usage.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="contributing.html" title="Contributing"
             >next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">smalr 1.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018John Beaulaurier.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.7.1.
    </div>
  </body>
</html>